### PR TITLE
Update vmsnapshot error phase in case of volumesnapshot errors which prevent the snapshot from being ready

### PR DESCRIPTION
**Pull request by ShellyKa13**
**What this PR does / why we need it**:
When trying to take VM snapshot of a VM, when one of the disk volumes is in Terminating
 state, if one try to take a snapshot in this state, the process will "silienty" fails on timeout with 
"snapshot deadline exceeded" 
This PR changes the phase to be "Errored" with conditions and error explaining the issue

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # Jira-ticket: https://issues.redhat.com/browse/CNV-35452

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
Expose volumesnapshot error in vmsnapshot object
```

### CODE REVIEW

**Comment by alicefr** [comment]
@ShellyKa13 the code looks good to me. Do you think the functional test is necessary?

**Comment by ShellyKa13** [comment]
@alicefr I was thinking If I should add functest or not, cause I do think the UT cover it pretty good.. I dont mind deleting it

**Comment by alicefr** [comment]
> @alicefr I was thinking If I should add functest or not, cause I do think the UT cover it pretty good.. I dont mind deleting it

Yeah, I have the same impression :)

**Comment by ShellyKa13** [comment]
@alicefr I do think that there is a value in functest since in case of a change in behavior in k8s for example then this test can catch problematic changes comparing to UT, wdyt?

**Comment by alicefr** [comment]
> @alicefr I do think that there is a value in functest since in case of a change in behavior in k8s for example then this test can catch problematic changes comparing to UT, wdyt?

I'm fine if you think this can be useful

File: staging/src/kubevirt.io/api/snapshot/v1alpha1/types.go [hunk 1]
```diff
@@ -91,6 +91,7 @@ const (
 	Succeeded  VirtualMachineSnapshotPhase = "Succeeded"
 	Failed     VirtualMachineSnapshotPhase = "Failed"
 	Deleting   VirtualMachineSnapshotPhase = "Deleting"
+	Errored    VirtualMachineSnapshotPhase = "ErrorState"
```
Comment (EdDev) [inline staging/src/kubevirt.io/api/snapshot/v1alpha1/types.go:94]: Unfortunately, this is problematic in terms of backward compatibility with older clients.

We have a combination of two points that makes it problematic:
- The use of a `phase` is discouraged [1] due to its state machine interpretation. Older clients are not aware of new added states, therefore they may handle new states as an exception path. While it may be ok in some cases, the general default is to avoid it and use conditions instead.
- When discussing the use of API enums [2], similar reasoning may apply even for a different usage than the one mentioned before.  If the value is expected to be interpreted as informative with no ties to a specific state machine, then it is safe for use.

[1] https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md?plain=1#L495
[2] https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-enum-value-in-existing-field

I would recommend to leave this information as part of the condition and instruct clients to query for the condition in order to check if there is an error state.
Avoid changing the existing phases and possibly mark it for deprecation (but not removal) to avoid clients for depending on it.

**Comment by EdDev** [comment]
@alaypatel07 FYI

/sig api

File: staging/src/kubevirt.io/api/snapshot/v1alpha1/types.go (hunk 1 above)
Comment (ShellyKa13) [reply staging/src/kubevirt.io/api/snapshot/v1alpha1/types.go]: @EdDev ok after some thought I agree in this case that the added value is no great since we have the conditions and the 'Error' in the status to reflect that. I added another commit to remove, please take a look

File: pkg/storage/snapshot/snapshot.go [hunk 2]
```diff
@@ -92,8 +92,7 @@ func vmSnapshotSucceeded(vmSnapshot *snapshotv1.VirtualMachineSnapshot) bool {
 }
 
 func vmSnapshotProgressing(vmSnapshot *snapshotv1.VirtualMachineSnapshot) bool {
-	return vmSnapshotError(vmSnapshot) == nil && !VmSnapshotReady(vmSnapshot) &&
```
[the remark this answers is not in the document]
Comment (ShellyKa13) [reply pkg/storage/snapshot/snapshot.go:95]: With the error being transient and possibly to have one and then it is gone the vmsnapshot can still be in progress and have error != nil.
`vmSnapshotProgressing` is used in several places for example `canUnlockSource` and I dont think we need to allow unlock source if error != nil, we should if the phase is actually different then in progress.
In this check in case of error we will return false: https://github.com/kubevirt/kubevirt/blob/02621d08b868c50288e803858936e7cf4fba3400/pkg/storage/snapshot/snapshot.go#L175
and then in the else if can unlock we will return true: https://github.com/kubevirt/kubevirt/blob/02621d08b868c50288e803858936e7cf4fba3400/pkg/storage/snapshot/snapshot.go#L195
so basically I want to continue vmsnapshot and stay in progress in case of error.

File: staging/src/kubevirt.io/api/snapshot/v1alpha1/types.go (hunk 1 above)
Comment (EdDev) [reply staging/src/kubevirt.io/api/snapshot/v1alpha1/types.go]: Thank you.

It will be nice if the last commit will be squashed in this first commit, to avoid adding and removing (it is a bit confusing).

**Review by EdDev — COMMENTED** [review]
The API change has been adjusted.

/unhold

File: tests/storage/snapshot.go [hunk 3]
```diff
@@ -1306,6 +1306,49 @@ var _ = SIGDescribe("VirtualMachineSnapshot Tests", func() {
 
 				Expect(snapshot.Status.CreationTime).To(BeNil())
 			})
+
+			It("vmsnapshot should update error if vmsnapshotcontent is unready to use and error", func() {
+				vm = tests.StartVMAndExpectRunning(virtClient, vm)
+				// Delete DV and wait pvc get deletionTimestamp
+				// when pvc is deleting snapshot is not possible
```
Comment (alromeros) [inline tests/storage/snapshot.go:1313]: Do we need the PVC to be there when we create the snapshot? What if the PVC is already deleted?

File: pkg/storage/snapshot/snapshot_test.go [hunk 4]
```diff
@@ -1111,6 +1140,111 @@ var _ = Describe("Snapshot controlleer", func() {
 				controller.processVMSnapshotWorkItem()
 			})
 
+			It("should update VirtualMachineSnapshot error when VirtualMachineSnapshotContent error", func() {
```
Comment (alromeros) [inline pkg/storage/snapshot/snapshot_test.go:1143]: Really good test coverage, kudos

File: pkg/storage/snapshot/snapshot.go [hunk 5]
```diff
@@ -706,6 +708,9 @@ func (ctrl *VMSnapshotController) updateSnapshotStatus(vmSnapshot *snapshotv1.Vi
 		vmSnapshotCpy.Status.Phase = snapshotv1.Failed
 		updateSnapshotCondition(vmSnapshotCpy, newProgressingCondition(corev1.ConditionFalse, vmSnapshotDeadlineExceededError))
 		updateSnapshotCondition(vmSnapshotCpy, newFailureCondition(corev1.ConditionTrue, vmSnapshotDeadlineExceededError))
+	} else if vmSnapshotError(vmSnapshotCpy) != nil {
```
Comment (alromeros) [inline pkg/storage/snapshot/snapshot.go:711]: Just out of curiosity, why move this?

**Review by alromeros — COMMENTED** [review]
Good job! looks good to me but still have a question about the functional test.

File: tests/storage/snapshot.go (hunk 3 above)
Comment (ShellyKa13) [reply tests/storage/snapshot.go:1313]: pvc cant be deleted as long as it it mounted in the vm, once the vm is down it can and will be deleted. We need the pvc to be in deleting phase cause if it is already deleted then we have no problem since there will not be a volume snapshot done for it.

File: pkg/storage/snapshot/snapshot_test.go (hunk 4 above)
Comment (ShellyKa13) [reply pkg/storage/snapshot/snapshot_test.go:1143]: Thanks :)

File: pkg/storage/snapshot/snapshot.go (hunk 5 above)
Comment (ShellyKa13) [reply pkg/storage/snapshot/snapshot.go:711]: VM can be in phase progressing while also have an Error, I want this if to happen instead of the progressing if.
